### ISSUE TITLE
Improve the FORCE_QUEUE process in Helix-machines

### ISSUE THREAD

**Issue by riarenas** [dotnet/arcade#13212]
- [ ] This issue is blocking 
- [ ] This issue is causing unreasonable pain

By default, PRs to helix-machines will only attempt to generate images for a single windows and a single linux queue. This is often not enough testing for some of the changes to machine configurations, which have different subtleties depending on the OS and version.

We have the [FORCE_QUEUE](https://dnceng.visualstudio.com/internal/_git/dotnet-helix-machines?path=/README.md&version=GBmain&line=47&lineEnd=48&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents) feature that allows to generate specific images during the PR process. However, this feature is sometimes not used, and we end up having to revert changes that fail in specific OSes that weren't tested during the PR.

We should look into ways of making the FORCE_QUEUE feature more visible, so that both PR authors and code reviewers are more easily aware of any gaps in their testing.

### Release Note Category

- [ ] Feature changes/additions 
- [ ] Bug fixes
- [x] Internal Infrastructure Improvements

### Release Note Description

For internal use only&hellip;

We have removed default queues and tightened the `FORCE_QUEUE` logic. Specifically:

- build only `FORCE_QUEUE` images in PR builds
  - avoid complications like checking for other queues in the same OS group
  - those checks previously chose the first Linux and Windows definitions found as defaults
- require at least two `FORCE_QUEUE`s in PR and manual builds
  - named queues must exist and require distinct images
- support `BUILD_EXISTING_IMAGES` in commit descriptions to force rebuilds
- add words to the PR template (when targeting 'main') about how to use `FORCE_QUEUE` and `BUILD_EXISTING_IMAGES`
- refactor some of the `CreateCustomImage` code; change logging and comments

**Comment by chcosta** [comment]
@riarenas - just a note, so we don't forget our hallway conversation : consider removing default queues (just make them suggestions) so people are forced to be intentional about which queues are running (require at least one queue [maybe more]) be defined.

**Comment by dougbu** [comment]
Not sure if this is covered in the existing ideas but `FORCE_QUEUE` is buggy in itself. I have a draft PR (<https://dev.azure.com/dnceng/internal/_git/dotnet-helix-machines/pullrequest/31499>) which includes two commits. One includes the following:
``` text
FORCE_QUEUE redhat.7.amd64
FORCE_QUEUE sles.12.amd64
FORCE_QUEUE ubuntu.1604.amd64
FORCE_QUEUE osx.1015.amd64
FORCE_QUEUE windows.7.amd64
FORCE_QUEUE windows.server.amd64.vs2017
```

A recent test build (<https://dev.azure.com/dnceng/internal/_build/results?buildId=2194461&view=results>) said it skipped building sles.12.amd64, ubuntu.16.04.amd64, and osx.1015.amd64 images but, in one case, actually created the image.
``` text
2023-06-06 20:43:57Z Skipping redhat.7.amd64, redhat.7.amd64.open, redhat.7.amd64.arcade, redhat.7.amd64.open.arcade, redhat.7.amd64.rt, redhat.7.amd64.open.rt, redhat.7.amd64.svc, redhat.7.amd64.open.svc with id -r7ZK59gyzs1m-e2s5iKkwsVzeaYjEQjHq9BWBIGtds because we decided centos.7.amd64, centos.7.amd64.open, centos.7.amd64.rt, centos.7.amd64.open.rt, centos.7.amd64.svc, centos.7.amd64.open.svc was good enough
2023-06-06 20:43:57Z Need to generate new image for redhat.7.amd64, redhat.7.amd64.open, redhat.7.amd64.arcade, redhat.7.amd64.open.arcade, redhat.7.amd64.rt, redhat.7.amd64.open.rt, redhat.7.amd64.svc, redhat.7.amd64.open.svc with id -r7ZK59gyzs1m-e2s5iKkwsVzeaYjEQjHq9BWBIGtds
2023-06-06 20:43:57Z Generating image for redhat.7.amd64, redhat.7.amd64.open, redhat.7.amd64.arcade, redhat.7.amd64.open.arcade, redhat.7.amd64.rt, redhat.7.amd64.open.rt, redhat.7.amd64.svc, redhat.7.amd64.open.svc
```

It's reasonable that the `FORCE_QUEUE osx.1015.amd64` part was ignored because machines in that queue are on-premise. However the logging was incorrect for redhat.7.amd64 and the sles.12.amd64 and ubuntu.1604.amd64 images should have been generated.

I suspect this is an order-of-operations problem. `ShouldCreateImage(...)` is called even for `FORCE_QUEUE` images (causing some incorrect logging) and `NeedToCreateImage(...)` returns `false` w/o logging in one case (likely causing `FORCE_QUEUE` to be ignored).

Bottom line, may be worthwhile simplifying the `FORCE_QUEUE` logic to unconditionally build and deploy them except in the unusual case that deployment would overwrite a consistent image already in the PR's scale set).

**Comment by dougbu** [comment]
Has any work in this area been funded❔

**Comment by dougbu** [comment]
> we instead parse the PR description for a specific section that would have all the queues to be forced

the problem here is that the PR description becomes the default text of the merged commit. that adds useless information to the 'main' and `production` histories.

it's relatively easy to rewrite history and `--force` when pushing but I'm willing to be convinced here.

**Comment by riarenas** [comment]
I think it's a good thing that it will show up in the commit information. That way we will know which queues were tested on each PR.

**Comment by premun** [comment]
I agree with @dougbu that we should remove any magic from the behaviour. If I `FORCE_QUEUE` some queues, I should expect them to get built. The logic around the choice of images to build is needlessly complicated and written in a needlessly complicated way.

Ad the problem of having FORCE_QUEUE in description/commit message, maybe we could utilize this "Tags" section and make people add tags to the PR? One per queue they want?

![image](https://github.com/dotnet/arcade/assets/7013027/0262aa08-ef47-4dc2-838a-5bb66f5f2734)

**Comment by dougbu** [comment]
my attempt to use `FORCE_QUEUE` is stymied by the complications @premun mentioned. I'm taking this because I need a fix in my draft PR to get the images I need built.

**Comment by dougbu** [comment]
this is complete but I'm not closing it until we do a build of main. [!32013](https://dev.azure.com/dnceng/internal/_git/dotnet-helix-machines/pullrequest/32013) should have no impact on official builds.

**Comment by dougbu** [comment]
Builds of main are no longer showing ill effects of the fix. Closing this issue. If necessary, we'll track the latest build break (which might just require a few more retries0 in a new issue.
